### PR TITLE
ddl: abstract job submitter & move some fields from ddlCtx to jobContext

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #54436

Problem Summary:

### What changed and how does it work?
- abstract job submitter
- move some fields to jobContext from ddlCtx
- simplify and move notifyNewJobSubmitted from executor to job submitter. keep value of addingDDLJobNotifyKey fixed to '0'
- move `init job-done-channel` to `GenGIDAndInsertJobsWithRetry` to make sure the channel exist before we start run the job.
- replace `SetBinlogClient` with a function variable
- some renaming
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/executor.go [hunk 1]
```diff
@@ -6306,8 +6306,6 @@ func (e *executor) DoDDLJobWrapper(ctx sessionctx.Context, jobW *JobWrapper) err
 	sessVars.StmtCtx.IsDDLJobInQueue = true
 
 	ddlAction := job.Type
-	// Notice worker that we push a new job and wait the job done.
-	e.notifyNewJobSubmitted(e.ddlJobNotifyCh, addingDDLJobNotifyKey, jobID, ddlAction.String())
```
Comment (D3Hunter) [inline pkg/ddl/executor.go:6310]: moved to job submitter

File: pkg/ddl/executor_test.go [hunk 2]
```diff
@@ -259,325 +250,3 @@ func TestCreateDropCreateTable(t *testing.T) {
 	require.Less(t, create0TS, dropTS, "first create should finish before drop")
 	require.Less(t, dropTS, create1TS, "second create should finish after drop")
 }
-
-func getGlobalID(ctx context.Context, t *testing.T, store kv.Storage) int64 {
```
Comment (D3Hunter) [inline pkg/ddl/executor_test.go:263]: moved to job_submitter_test.go

File: pkg/ddl/job_submitter.go [hunk 3]
```diff
@@ -320,26 +349,19 @@ func (d *ddl) addBatchDDLJobs2Table(jobWs []*JobWrapper) error {
 
 	se.GetSessionVars().SetDiskFullOpt(kvrpcpb.DiskFullOpt_AllowedOnAlmostFull)
 	ddlSe := sess.NewSession(se)
-	if err = GenGIDAndInsertJobsWithRetry(ctx, ddlSe, jobWs); err != nil {
+	if err = s.GenGIDAndInsertJobsWithRetry(ctx, ddlSe, jobWs); err != nil {
 		return errors.Trace(err)
 	}
-	for _, jobW := range jobWs {
```
Comment (D3Hunter) [inline pkg/ddl/job_submitter.go:326]: moved into the gen&insert transaction

File: pkg/ddl/job_submitter.go [hunk 4]
```diff
@@ -637,33 +674,21 @@ func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
 	return nil
 }
 
-func (e *executor) notifyNewJobSubmitted(ch chan struct{}, etcdPath string, jobID int64, jobType string) {
-	// If the workers don't run, we needn't notify workers.
-	// TODO: It does not affect informing the backfill worker.
-	if !config.GetGlobalConfig().Instance.TiDBEnableDDL.Load() {
```
Comment (D3Hunter) [inline pkg/ddl/job_submitter.go:643]: if `!config.GetGlobalConfig().Instance.TiDBEnableDDL.Load()` is true, then it's not the owner, we need notify by etcd

File: pkg/ddl/job_submitter.go [hunk 5]
```diff
@@ -637,33 +674,21 @@ func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
...
-	if e.etcdCli == nil {
+func (s *JobSubmitter) notifyNewJobByEtcd() {
+	if s.etcdCli == nil {
 		return
 	}
 
-	jobIDStr := strconv.FormatInt(jobID, 10)
-	timeStart := time.Now()
-	err := ddlutil.PutKVToEtcd(e.ctx, e.etcdCli, 1, etcdPath, jobIDStr)
+	err := ddlutil.PutKVToEtcd(s.ctx, s.etcdCli, 1, addingDDLJobNotifyKey, "0")
```
Comment (D3Hunter) [inline pkg/ddl/job_submitter.go:690]: value doesn't matter

File: pkg/ddl/job_submitter_test.go
Comment (D3Hunter) [inline pkg/ddl/job_submitter_test.go:1]: only TestGenGIDAndInsertJobsWithRetryOnErr is new added

File: pkg/ddl/job_worker.go [hunk 6]
```diff
@@ -981,25 +995,21 @@ func toTError(err error) *terror.Error {
...
+// to reload info schema, and waits for all servers' schema or MDL synced.
+func updateGlobalVersionAndWaitSynced(jobCtx *jobContext, latestSchemaVersion int64, job *model.Job) error {
 	if !job.IsRunning() && !job.IsRollingback() && !job.IsDone() && !job.IsRollbackDone() {
 		return nil
 	}
 
-	timeStart := time.Now()
 	var err error
-	defer func() {
-		metrics.DDLWorkerHistogram.WithLabelValues(metrics.WorkerWaitSchemaChanged, job.Type.String(), metrics.RetLabel(err)).Observe(time.Since(timeStart).Seconds())
```
Comment (D3Hunter) [inline pkg/ddl/job_worker.go:994]: moved to waitVersionSynced

**Comment by D3Hunter** [comment]
@lance6716

**Review by lance6716 — APPROVED** [review]

**Pull request merged**